doc/design/design_cluster_job_runner.md

```diff
@@ -0,0 +1,126 @@
+
+1. The local job may cause the SQLFlow server resource insufficient when there are too much SQL jobs.
+1. Sometimes, the SQL job takes too much time and the gRPC calls timeout.
+1. The SQLFlow server is not High-Available, if an SQLFlow server instance failed, the jobs on this instance are failed.
+
+In this design, we propose to implement the **Cluster Job Runner** to solve the above problems.
+
+The cluster job runner would launch the job on Kubernetes or Yarn cluster instead of on the host. The SQLflow client can
+check the job status in a polling manner instead of a long live connection. We recommend using the cluster job runner in the production environment.
+
+## High-Level Design
+
+For the most submitter program, it can run as the local mode or distributed mode. For the local mode, SQLFlow would run
+the submitter program as a local process on the host; For the distributed model, the submitter program would submit a distributed Job to the cluster. For the two modes, the behavior of SQLFlow is different in both local job runner and cluster job runner:
+
+Job Runner| local mode | distributed mode
+-- | -- | --
+Local | launch a process on the host with blocking| submite a job to cluster with blocking
+Cluster| launch a Kubernetes Pod with no-blocing| submite a job to cluster with no-blocking
+
+The cluster job runner workflow is as follows:
+
+<img src="figures/cluster_job_runner.png">
+
+1. SQLFlow client sends the SQL statement via a gRPC call to the SQLFlow server. The local job goto **2** and the distributed
+job go to **3**.
+2. If the SQL statement implies a local job, e.g. local TensorFlow Job and local XGBoost job.
+    1. SQLFLow server would launch a Pod on Kubernetes cluster via Kubernetes API.
+    1. SQLFLow server can fetch the job tracker URL from the api-server.
+    1. The SQLFlow server returned the job tracker URL to the client.
+3. If the SQL statement implies a distributed job, e.g. distributed Tensorflow Job, distributed ALPS job, or other distributed
+machine learning job.
+    1. SQLFlow server would call the distributed AI framework API to launch a distributed Job on a cluster(Kubernetes/Yarn),
+    e.g. EDL API or ALPS API.
+    1. SQLFLow server can fetch the Job tracer URL from the distributed AI framework API.
+    1. SQLFlow server returns the job tracer URL to the client.
+4. The SQLFLow client would send the job tracer URL to the SQLFlow server to request the task status in a polling manner until
+the returned task status is **COMPLETED** or **FAILED**.
+
+## Proposal Details
+
+1. Add a new gRPC interface to check the job status.
+
+    ``` protobuf
+    service SQLFlow {
+      /**
+      if sqlflow_job_runner == "local":
+          sqlflow.Run(sql...)
+      elif sqlflow_job_runner == "cluster":
+          res = sqlflow.Run(sql...)
+          is_finised = False
+          while !is_finished:
+              is_finised = sqlflow.IsFinished(res.session.job_tracker_url)
```

Comment:
So if this is implemented, does ElasticDL still need to provide a blocking/sync API? Or maybe ElasticDL just needs to provide an API that can be called to check the job status? See https://github.com/sql-machine-learning/elasticdl/issues/1285 for related issue. cc @tonyyang-svail

Reply:
I believe "an API that can be called to check the job status" is better since a training job could last weeks while the long connection can timeout.

Reply:
Maybe implement a blocking API that checks the EDL job in a polling manner is better.

Reply:
@Yancey1989 This proposal handles the polling manner already though. ElasticDL only needs to provide the API that can be called each time to check the status.